apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/AllocateAndAddFloatingIpToNode.java

```diff
@@ -156,8 +215,57 @@ public boolean apply(FloatingIP arg0) {
       return Optional.fromNullable(ip);
    }
 
+   private Optional<org.jclouds.openstack.neutron.v2.domain.FloatingIP> tryFindExistingFloatingIp(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi, final String availabilityZone) {
+      Optional<org.jclouds.openstack.neutron.v2.domain.FloatingIP> floatingIPOptional = neutronFloatingApi.list().concat().firstMatch(new Predicate<org.jclouds.openstack.neutron.v2.domain.FloatingIP>() {
+         @Override
+         public boolean apply(@Nullable org.jclouds.openstack.neutron.v2.domain.FloatingIP input) {
+            return input.getPortId() == null && input.getAvailabilityZone().equals(availabilityZone);
+         }
+      });
+      return floatingIPOptional;
+   }
+
+   private org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIpUsingNeutron(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi, NodeMetadata node, final String availabilityZone) {
+      String regionId = node.getLocation().getParent().getId();
+
+      Optional<Network> networkOptional = getNetworkApi(regionId).list().concat().firstMatch(new Predicate<Network>() {
+         @Override
+         public boolean apply(@Nullable Network input) {
+            return input.getExternal() && input.getAvailabilityZone().equals(availabilityZone);
+         }
+      });
+      if (!networkOptional.isPresent())
+         throw new InsufficientResourcesException("Failed to find a suitable external network.");
+
+      org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIP = org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP
+              .createBuilder(networkOptional.get().getId())
+              .availabilityZone(networkOptional.get().getAvailabilityZone())
+              .build();
+
+      org.jclouds.openstack.neutron.v2.domain.FloatingIP floatingIP = neutronFloatingApi.create((org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP) createFloatingIP);
+      floatingIpCache.asMap().putIfAbsent(RegionAndId.fromSlashEncoded(node.getId()), ImmutableList.of(FloatingIpForServer.create(RegionAndId.fromSlashEncoded(node.getId()), floatingIP.getId(), floatingIP.getFloatingIpAddress())));
```

Comment:
Why put if absent? Shouldn't we override the old value, if any?

Reply:
hmm, makes sense, although I was using the same approach used for Nova. Should I change line 141 as well?

Reply:
I'd say so?